src/common/heap_profiling/src/profiler.rs

```diff
@@ -44,79 +44,63 @@ impl HeapProfiler {
-                .to_owned();
-            let file_path_c = CString::new(file_path).expect("0 byte in file path");
+        let file_path = Path::new(&self.config.dir)
+            .join(&file_name)
+            .to_str()
+            .expect("file path is not valid utf8")
+            .to_owned();
+        let file_path_c = CString::new(file_path).expect("0 byte in file path");
 
-            // FIXME(yuhao): `unsafe` here because `jemalloc_dump_mib.write` requires static lifetime
-            if let Err(e) = self
-                .jemalloc_dump_mib
-                .write(unsafe { &*(file_path_c.as_c_str() as *const _) })
-            {
-                tracing::warn!("Auto Jemalloc dump heap file failed! {:?}", e);
-            } else {
-                tracing::info!("Successfully dumped heap profile to {}", file_name);
-            }
+        // FIXME(yuhao): `unsafe` here because `jemalloc_dump_mib.write` requires static lifetime
+        if let Err(e) = self
+            .jemalloc_dump_mib
+            .write(unsafe { &*(file_path_c.as_c_str() as *const _) })
+        {
+            tracing::warn!("Auto Jemalloc dump heap file failed! {:?}", e);
+        } else {
+            tracing::info!("Successfully dumped heap profile to {}", file_name);
         }
     }
 
-    fn advance_jemalloc_epoch(&self, prev_jemalloc_allocated_bytes: usize) -> usize {
-        if let Err(e) = self.jemalloc_epoch_mib.advance() {
-            tracing::warn!("Jemalloc epoch advance failed! {:?}", e);
+    /// Start the daemon task of auto heap profiling.
+    pub fn start(self) {
+        if !self.config.enable_auto {
+            tracing::info!("Auto memory dump is disabled with enable_auto=false");
+            return;
         }
 
-        self.jemalloc_allocated_mib.read().unwrap_or_else(|e| {
-            tracing::warn!("Jemalloc read allocated failed! {:?}", e);
-            prev_jemalloc_allocated_bytes
-        })
-    }
-
-    pub fn start(self) {
         static START: Once = Once::new();
         START.call_once(|| {
             fs::create_dir_all(&self.config.dir).unwrap();
             tokio::spawn(async move {
                 let mut interval = time::interval(Duration::from_millis(500));
-                let mut prev_jemalloc_allocated_bytes = 0;
+                let mut prev_used_memory_bytes = 0;
                 loop {
                     interval.tick().await;
-                    let jemalloc_allocated_bytes =
-                        self.advance_jemalloc_epoch(prev_jemalloc_allocated_bytes);
-                    self.dump_heap_prof(jemalloc_allocated_bytes, prev_jemalloc_allocated_bytes);
-                    prev_jemalloc_allocated_bytes = jemalloc_allocated_bytes;
+                    let cur_used_memory_bytes = resource_util::memory::total_memory_used_bytes();
+
+                    // Dump heap profile when memory usage is crossing the threshold.
+                    if self.opt_prof
```

Comment:
nit: why not also put `self.opt_prof` at the beginning?

Reply:
Sounds good